### PR TITLE
feat: table route for metric engine

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Now metric engine table can be created. Basic insertion is supported:

```text
mysql> create table physical_table(ts timestamp time index) engine = metric with (physical_metric_table = true);
Query OK, 0 rows affected (0.03 sec)

mysql> select * from information_schema.tables where engine = 'metric';
+---------------+--------------+----------------+------------+----------+--------+
| table_catalog | table_schema | table_name     | table_type | table_id | engine |
+---------------+--------------+----------------+------------+----------+--------+
| greptime      | public       | physical_table | BASE TABLE |     1044 | metric |
+---------------+--------------+----------------+------------+----------+--------+
1 row in set (0.48 sec)

mysql> create table l1(ts timestamp time index, i int) engine = metric with (on_physical_table = 'physical_table');
Query OK, 0 rows affected (0.05 sec)

mysql> insert into l1 values (1,1);
Query OK, 1 row affected (0.01 sec)

mysql> select * from l1;
+----------------------------+------+
| ts                         | i    |
+----------------------------+------+
| 1970-01-01 00:00:00.001000 |    1 |
+----------------------------+------+
1 row in set (0.03 sec)

mysql> create table l2(ts timestamp time index, s string) engine = metric with (on_physical_table = 'physical_table');
Query OK, 0 rows affected (0.05 sec)

mysql> insert into l2 values(1,"a");
Query OK, 1 row affected (0.00 sec)
```

Basic query is also supported:

```text
mysql> select * from l2;
+----------------------------+------+
| ts                         | s    |
+----------------------------+------+
| 1970-01-01 00:00:00.001000 | NULL | <-- Not expected line, might be related to some unfinished metric engine works though. Ignore it now.
| 1970-01-01 00:00:00.001000 | a    |
+----------------------------+------+
2 rows in set (0.01 sec)

```

Will add more tests in the next PR.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/2864

### CODE REVIEW

**Comment by fengjiachun** [comment]
@waynexia PTAL

**Review by fengjiachun — APPROVED** [review]
LGTM

**Review by fengjiachun — CHANGES_REQUESTED** [review]
I conducted a test: after restarting, the two logical tables l1 and l2 could not be found.

**Comment by MichaelScofield** [comment]
@fengjiachun PTAL

File: src/datanode/src/region_server.rs [hunk 1]
```diff
@@ -505,16 +509,20 @@ impl RegionServerInner {
...
         engine: RegionEngineRef,
         region_change: RegionChange,
-    ) {
+    ) -> Result<()> {
         let engine_type = engine.name();
         match region_change {
             RegionChange::None => {}
             RegionChange::Register(_) => {
+                if engine_type == METRIC_ENGINE_NAME {
+                    self.register_logical_regions(&engine, region_id).await?;
```
Comment (fengjiachun) [inline src/datanode/src/region_server.rs:523]: Create a logical table is also a `RegionChange::Register` event

File: src/partition/src/metrics.rs
Comment (waynexia) [inline src/partition/src/metrics.rs:1]: Can we remove this file and mod?

File: src/datanode/src/region_server.rs [hunk 2]
```diff
@@ -528,6 +536,37 @@ impl RegionServerInner {
                 self.event_listener.on_region_deregistered(region_id);
             }
         }
+        Ok(())
+    }
+
+    async fn register_logical_regions(
```
Comment (waynexia) [inline src/datanode/src/region_server.rs:542]: I'm wondering if it's possible to merge this logic with region open action, but it also works now. We can consider this problem later

**Review by waynexia — APPROVED** [review]
This LGTM in general :+1:

File: src/common/meta/src/ddl/table_meta.rs [hunk 3]
```diff
@@ -115,8 +123,31 @@ impl TableMetadataAllocator {
     ) -> Result<TableRouteValue> {
         let regions = task.partitions.len();
 
-        let table_route = if task.create_table.engine == METRIC_ENGINE {
-            TableRouteValue::Logical(LogicalTableRouteValue {})
+        let table_route = if task.create_table.engine == METRIC_ENGINE
+            && let Some(physical_table_name) = task
+                .create_table
+                .table_options
+                .get(LOGICAL_TABLE_METADATA_KEY)
```
Comment (WenyXu) [inline src/common/meta/src/ddl/table_meta.rs:130]: What if the `table_options` is missing `LOGICAL_TABLE_METADATA_KEY`? Should we change to following: 
```rust 
let table_route = if task.create_table.engine == METRIC_ENGINE {
            let physical_table_name = task
                .create_table
                .table_options
                .get(LOGICAL_TABLE_METADATA_KEY)
                .context(error::UnexpectedSnafu {})?;
        
```

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**